In [54]:
#Import required libraries
import keras #library for neural network
import pandas as pd #loading data in table form
import seaborn as sns #visualisation
import matplotlib.pyplot as plt #visualisation
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.preprocessing import normalize #machine learning algorithm library


In [55]:
#Reading data
data=pd.read_csv("https://raw.githubusercontent.com/kanadakurniawan/Belajar-Jaringan-Saraf-Tiruan/1d95438528ffade3bde5554b0402cf2c5c262ec9/dataset/Iris.csv")
print("Describing the data: ",data.describe())
print("Info of the data:",data.info())

Describing the data:                 Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count  150.000000     150.000000    150.000000     150.000000    150.000000
mean    75.500000       5.843333      3.054000       3.758667      1.198667
std     43.445368       0.828066      0.433594       1.764420      0.763161
min      1.000000       4.300000      2.000000       1.000000      0.100000
25%     38.250000       5.100000      2.800000       1.600000      0.300000
50%     75.500000       5.800000      3.000000       4.350000      1.300000
75%    112.750000       6.400000      3.300000       5.100000      1.800000
max    150.000000       7.900000      4.400000       6.900000      2.500000
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidt

In [56]:
print("10 first samples of the dataset:",data.head(10))
print("10 last samples of the dataset:",data.tail(10))

10 first samples of the dataset:    Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa
5   6            5.4           3.9            1.7           0.4  Iris-setosa
6   7            4.6           3.4            1.4           0.3  Iris-setosa
7   8            5.0           3.4            1.5           0.2  Iris-setosa
8   9            4.4           2.9            1.4           0.2  Iris-setosa
9  10            4.9           3.1            1.5           0.1  Iris-setosa
10 last samples of the dataset:       Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
140  141          

Pengkodean di bawah ini mengubah spesies ke dalam masing-masing kategori untuk dimasukkan ke dalam jaringan saraf

In [57]:
print(data["Species"].unique())

['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']


In [58]:
data.loc[data["Species"]=="Iris-setosa","Species"]=0
data.loc[data["Species"]=="Iris-versicolor","Species"]=1
data.loc[data["Species"]=="Iris-virginica","Species"]=2
print(data.head())

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm Species
0   1            5.1           3.5            1.4           0.2       0
1   2            4.9           3.0            1.4           0.2       0
2   3            4.7           3.2            1.3           0.2       0
3   4            4.6           3.1            1.5           0.2       0
4   5            5.0           3.6            1.4           0.2       0


In [59]:
data=data.iloc[np.random.permutation(len(data))]
print(data.head())

      Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm Species
16    17            5.4           3.9            1.3           0.4       0
75    76            6.6           3.0            4.4           1.4       1
143  144            6.8           3.2            5.9           2.3       2
86    87            6.7           3.1            4.7           1.5       1
5      6            5.4           3.9            1.7           0.4       0


Mengonversi data ke array numpy agar dapat diproses

In [60]:
X=data.iloc[:,1:5].values
y=data.iloc[:,5].values

print("Shape of X",X.shape)
print("Shape of y",y.shape)
print("Examples of X\n",X[:3])
print("Examples of y\n",y[:3])

Shape of X (150, 4)
Shape of y (150,)
Examples of X
 [[5.4 3.9 1.3 0.4]
 [6.6 3.  4.4 1.4]
 [6.8 3.2 5.9 2.3]]
Examples of y
 [0 1 2]


**Normalisasi**

Dapat dilihat dari atas bahwa fitur dataset pertama memiliki Panjang Kelopak 5.4 cm, Lebar Kelopak 3,9 cm, Panjang Kelopak 1.3 cm, dan Lebar Kelopak 0.4 cm. Namun, rentang dataset mungkin berbeda. Oleh karena itu, untuk mempertahankan akurasi yang baik, fitur setiap dataset harus dinormalisasi ke rentang 0-1 untuk diproses.

In [61]:
X_normalized=normalize(X,axis=0)
print("Examples of X_normalised\n",X_normalized[:3])

Examples of X_normalised
 [[0.07471338 0.10323929 0.02557886 0.02300599]
 [0.09131636 0.07941484 0.0865746  0.08052096]
 [0.09408352 0.08470916 0.11608866 0.13228444]]


In [62]:
#Membuat data pelatihan, pengujian, dan validasi
'''
80% -- train data
20% -- test data
'''
total_length=len(data)
train_length=int(0.8*total_length)
test_length=int(0.2*total_length)

X_train=X_normalized[:train_length]
X_test=X_normalized[train_length:]
y_train=y[:train_length]
y_test=y[train_length:]

print("Length of train set x:",X_train.shape[0],"y:",y_train.shape[0])
print("Length of test set x:",X_test.shape[0],"y:",y_test.shape[0])

Length of train set x: 120 y: 120
Length of test set x: 30 y: 30


In [63]:
# Neural network module
from keras.models import Sequential
from keras.layers import Dense, Activation, Dropout
# from keras.layers.normalization import BatchNormalization
from tensorflow.keras.utils import to_categorical

# Change the label to one hot vector
'''
[0] ---> [1 0 0]
[1] ---> [0 1 0]
[2] ---> [0 0 1]
'''
y_train = to_categorical(y_train, num_classes=3)
y_test = to_categorical(y_test, num_classes=3)

print("Shape of y_train", y_train.shape)
print("Shape of y_test", y_test.shape)


Shape of y_train (120, 3)
Shape of y_test (30, 3)


In [64]:
model=Sequential()
model.add(Dense(1000,input_dim=4,activation='relu'))
model.add(Dense(500,activation='relu'))
model.add(Dense(300,activation='relu'))
# model.add(Dropout(0.2))
model.add(Dense(3,activation='softmax'))
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [65]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_12 (Dense)                     │ (None, 1000)                │           5,000 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_13 (Dense)                     │ (None, 500)                 │         500,500 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_14 (Dense)                     │ (None, 300)                 │         150,300 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_15 (Dense)                     │ (None, 3)                   │             903 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 656,703 (2.51 MB)

 Trainable params: 656,703 (2.51 MB)

 Non-trainable params: 0 (0.00 B)

In [66]:
model.fit(X_train,y_train,validation_data=(X_test,y_test),batch_size=20,epochs=10,verbose=1)

Epoch 1/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy: 0.3179 - loss: 1.0901 - val_accuracy: 0.7667 - val_loss: 1.0527
Epoch 2/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.6993 - loss: 1.0292 - val_accuracy: 1.0000 - val_loss: 0.9462
Epoch 3/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7960 - loss: 0.9166 - val_accuracy: 0.9000 - val_loss: 0.7487
Epoch 4/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7727 - loss: 0.7466 - val_accuracy: 1.0000 - val_loss: 0.5336
Epoch 5/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step - accuracy: 0.8817 - loss: 0.5503 - val_accuracy: 0.8333 - val_loss: 0.4142
Epoch 6/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.9686 - loss: 0.3848 - val_accuracy: 0.9667 - val_loss: 0.2581
Epoch 7/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step - accuracy: 0.9649 - loss: 0.2743 - val_accuracy: 1.0000 - val_loss: 0.1376
Epoch 8/10
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step - accuracy: 0.9550 - loss: 0.2040 - val_accuracy: 1.0000 - val_loss: 0.1020


In [67]:
prediction=model.predict(X_test)
length=len(prediction)
y_label=np.argmax(y_test,axis=1)
predict_label=np.argmax(prediction,axis=1)

accuracy=np.sum(y_label==predict_label)/length * 100
print("Accuracy of the dataset",accuracy )

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 99ms/step
Accuracy of the dataset 96.66666666666667


Akurasi 96.67% dicapai dalam kumpulan data ini. Dapat dipastikan bahwa untuk setiap periode, jaringan saraf mencoba belajar dari fitur yang ada dan memprediksinya berdasarkan bobot dan biasnya. Untuk setiap periode, bobot dan bias diubah dengan mengurangi rasionya untuk mendapatkan akurasi yang lebih baik setiap saat.

In [68]:
print(X_test)
print(y_label)

[[0.08716562 0.06617903 0.09838022 0.10927845]
 [0.07056264 0.09000348 0.02951407 0.01150299]
 [0.10515217 0.07941484 0.12986189 0.12078145]
 [0.10930291 0.10059213 0.12592669 0.11502995]
 [0.07886413 0.06882619 0.06886616 0.05751497]
 [0.07609697 0.11118077 0.02754646 0.01150299]
 [0.07748055 0.06617903 0.07673657 0.06326647]
 [0.09823426 0.07941484 0.11608866 0.12078145]
 [0.06779548 0.082062   0.02951407 0.0057515 ]
 [0.08024771 0.07147336 0.10034783 0.10927845]
 [0.05949399 0.07941484 0.02164365 0.0057515 ]
 [0.0885492  0.08470916 0.10428304 0.13228444]
 [0.07471338 0.09794497 0.02951407 0.01150299]
 [0.06226115 0.06088471 0.02557886 0.01725449]
 [0.0885492  0.07676768 0.08460699 0.07476947]
 [0.0664119  0.09000348 0.03738448 0.01150299]
 [0.08163129 0.07941484 0.10034783 0.10352695]
 [0.07056264 0.10059213 0.02951407 0.01725449]
 [0.06917906 0.09000348 0.02951407 0.01150299]
 [0.06502832 0.08470916 0.02557886 0.01150299]
 [0.07609697 0.09265065 0.02557886 0.01150299]
 [0.07748055 

In [69]:
print(predict_label)

[2 0 2 2 1 0 1 2 0 2 0 2 0 0 1 0 2 0 0 0 0 1 2 2 2 0 2 1 0 2]
